model/riscv_sys_regs.sail

```diff
@@ -156,21 +156,24 @@ val sys_enable_next = {c: "sys_enable_next", ocaml: "Platform.enable_next", _: "
    unsetting C. If it returns true the write will have no effect. */
 val ext_veto_disable_C : unit -> bool effect {rreg}
 
-/* We currently only support dynamic changes for the C extension. */
 function legalize_misa(m : Misa, v : xlenbits) -> Misa = {
-  if   sys_enable_writable_misa ()
-  then { /* Handle modifications to C. */
-         let  v = Mk_Misa(v);
-         /* Suppress changing C if nextPC would become misaligned or an extension vetoes or C was disabled at boot (i.e. not supported). */
-         let m =
-           if   (v.C() == 0b0 & (nextPC[1] == bitone | ext_veto_disable_C())) | not(sys_enable_rvc())
-           then m
-           else update_C(m, v.C());
-         /* Handle updates for F/D. */
-         if   not(sys_enable_fdext()) | (v.D() == 0b1 & v.F() == 0b0)
-         then m
-         else update_D(update_F(m, v.F()), v.D())
-       }
+  /* If MISA is writable then suppress disabling C & keep misa unchanged if nextPC would become misaligned or an extension vetoes */
+  if   sys_enable_writable_misa() & (v.C() == 0b0 & (nextPC[1] == bitzero & not(ext_veto_disable())))
+  then {
+    let v = Mk_Misa(v);
```

Comment:
This needs hoisting

Reply:
(which you would have noticed had you tried to run make yourself; please don't push totally-untested code and waste our time)

Reply:
I don't understand this.

Reply:
v is xlenbits on entry to the function. In the if condition above you treat it like an misa bitfield. Therefore you must have `let v = Mk_Misa(v);` before it so that v is such a bitfield, i.e. you must hoist this line above the if just like you hoisted the condition.

Reply:
Or just look at the build error and observe that it is currently totally broken.

Reply:
@jrtc27 is spot-on in complaining about this.
A simple 'make' clearly shows the error:
```
Type error:
model/riscv_sys_regs.sail:161.7-100:
161 |  if   not(sys_enable_writable_misa()) | (v.C() == 0b0 & (nextPC[1] == bitone | ext_veto_disable()))
    |       ^-------------------------------------------------------------------------------------------^
    | No overloading for (operator |), tried:
    | * or_bool
    |    No overloading for (operator &), tried:
    |    * and_bool
    |       No overloading for (operator ==), tried:
    |       * eq_int
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bool
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_unit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_string
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bits
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_string
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_anything
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |    * and_vec
    |       No overloading for (operator ==), tried:
    |       * eq_int
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bool
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_unit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_string
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bits
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_string
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_anything
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |    * and_vec
    |       No overloading for (operator ==), tried:
    |       * eq_int
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bool
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_unit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_string
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bit
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_bits
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_string
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    |       * eq_anything
    |          No overloading for _mod_C, tried:
    |          * _get_Misa_C
    |             bitvector(64, dec) is not a subtype of Misa
    |          * _set_Misa_C
    |             Function _set_Misa_C applied to 1 args, expected 2 (2 explicit): register(Misa), bitvector(1, dec)
    | * or_vec
    |    Could not unify bitvector('n, dec) and bool(not('ex1703#))
    | * or_vec
    |    Could not unify bitvector('n, dec) and bool(not('ex1706#))
make: *** [Makefile:207: generated_definitions/ocaml/RV64/riscv.ml] Error 1
```